### ISSUE TITLE
plot_model not connecting layers between models correctly

### ISSUE THREAD

**Issue by will-cern** [tensorflow/tensorflow#42101]
Tested in tf 2.3.0

When building models from other models, the graph generated by `plot_model` does not create arrows between the correct outputs and inputs. See e.g. 

```
inn = tf.keras.Input((1,),name="in1")
d1 = tf.keras.layers.Dense(1, name="d1")(inn)
d2 = tf.keras.layers.Dense(1, name="d2")(inn)
m1 = tf.keras.Model(inputs=inn, outputs=[d1,d2], name="model1")
in2_1 = tf.keras.Input((1,),name="in2_1")
in2_2 = tf.keras.Input((1,),name="in2_2")
m2 = tf.keras.Model(inputs=[in2_1,in2_2],outputs=[in2_1 + in2_2],name="model2")

# combined model
in0 = tf.keras.Input((1,),name="in0")
m = tf.keras.Model(inputs=in0, outputs=m2(m1(in0)))
tf.keras.utils.plot_model(m, show_shapes=True, expand_nested=True)
```
outputs:
![image](https://user-images.githubusercontent.com/18280829/89551998-7bb23700-d803-11ea-961f-127eb8e71413.png)

As you can see, the first output of the first model isn't connected to the first output of the second model

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/1b993bb4c0c6c8cea3643816fe824650/42101-2-3.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/9d299572494a983430de30536ad0d8b2/42101-2-3.ipynb). Please find the attached gist. Thanks!

**Comment by tu-curious** [comment]
I have another issue with plot_model while using it for a double-nested model. I thought this might be the right place to post it. I will remove it if inappropriate. When I create a model_AE from 2 nested models (encoder,decoder), plot_model plots them correctly. But when I further nest model_AE with model_HR into the final end2end model_e2e, the plotted model is incorrect. The arrow goes from input of model_AE to input of model_HR but it should go from output of model_AE to input of model_HR. I have attached the code to reproduce it below and also the generated figure. Tried with Anaconda on Mac (tensorflow 2.0) and Windows (tensorflow 2.1). Any help is appreciated. Thanks.

    import tensorflow as tf
    from tensorflow.keras import layers
    from tensorflow import keras
    
    HR_win_len=200
    in_shape=(400,)
    latent_shape=(16,)
    
    #Simple AE model
    encoder_in = layers.Input(shape=in_shape,name="encoder_in")
    x = layers.Dense(256,activation='relu')(encoder_in)
    encoder_out = layers.Dense(latent_shape[0],activation='relu')(x)
    encoder = keras.Model(encoder_in, encoder_out, name="encoder")
    
    decoder_in = layers.Input(shape=latent_shape,name="decoder_in")
    x = layers.Dense(256,activation='relu')(decoder_in)
    decoder_out = layers.Dense(in_shape[0])(x)
    decoder = keras.Model(decoder_in, decoder_out, name="decoder")
    
    #Connect encoder and decoder and create model_AE
    AE_in = keras.Input(shape=in_shape, name="AE_in")
    sig_hat = decoder(encoder(AE_in))
    model_AE = keras.Model(AE_in, sig_hat, name="AE")
    
    #Create model_HR
    expand_dims = layers.Lambda(lambda x: tf.expand_dims(x,axis=-1), 
                                    name='expand_dims')
    rnn = layers.GRU(64, return_sequences=True, return_state=True)
    sig_in = layers.Input(shape=in_shape)
    x = expand_dims(sig_in)
    _, final_state=rnn(x[:,:HR_win_len,:]) #warm-up RNN
    rnn_out, _ = rnn(x[:,HR_win_len:,:],initial_state=final_state)
    HR_hat=layers.Conv1D(filters=1,kernel_size=1, strides=1,padding='same',
                         activation=None,name='Conv_{}'.format(1))(rnn_out)
    HR_hat=layers.Flatten()(HR_hat)
    model_HR = keras.Model(sig_in, HR_hat, name="HR")
    
    #Connect model_AE and model_HR and create model_e2e
    e2e_in = keras.Input(shape=in_shape, name="e2e_in")
    e2e_out=model_HR(model_AE(e2e_in))
    model_e2e = keras.Model(e2e_in, e2e_out, name="e2e")
    
    tf.keras.utils.plot_model(model_e2e,to_file='e2e.png', 
    dpi=200, show_shapes=True, show_layer_names=True, expand_nested=True)
[![enter image description here][1]][1]


  [1]: https://i.stack.imgur.com/y5TzS.png

**Comment by Doekeb** [comment]
Here to confirm that this is happening for me as well on Tensorflow 2.4.1. Here's MWE:
```
from tensorflow import keras

a = keras.Input(shape=(1,), name="a")
b = keras.Input(shape=(1,), name="b")
c = keras.layers.Concatenate(name="c")([a, b])
inner = keras.Model(inputs=[a, b], outputs=c, name="inner")

x = keras.Input(shape=(1,), name="x")
y = keras.Input(shape=(1,), name="y")
z = inner([x, y])
outer = keras.Model(inputs=[x, y], outputs=z, name="outer")

keras.utils.plot_model(outer, show_shapes=True, expand_nested=True)
```

![model](https://user-images.githubusercontent.com/13689478/109202594-2757b000-7771-11eb-9dd4-a39a34305455.png)

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue in TF v2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/6a8bfaddbd127216f327724b2d774658/untitled103.ipynb)..Thanks !

**Comment by sanatmpa1** [comment]
Issue still persists in TF `2.7` and TF-nightly `2.8.0-dev20211120`. Please find the [gist here](https://colab.sandbox.google.com/gist/sanatmpa1/39abf88d40f910bb1aae42640aae283b/untitled103.ipynb).

**Comment by jvishnuvardhan** [comment]
@will-cern Agree. This is still an issue. Is it possible for you to open this issue in keras-team/keras repo as the development of Keras moved to that repo.

Keras development moved to another repository to focus on only keras. Could you please repost this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
